src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+
+    let store =
+        ConcurrentDictionary<'Key, CachedEntity<'Key, 'Value>>(Environment.ProcessorCount, totalCapacity)
+
+    let evictionQueue = LinkedList<CachedEntity<'Key, 'Value>>()
+
+    let addToEvictionQueue (entity: CachedEntity<'Key, 'Value>) =
+        lock evictionQueue <| fun () -> evictionQueue.AddLast(entity.Node)
+
+    // Only LRU currrently. We can add other strategies when needed.
+    let updateEvictionQueue (entity: CachedEntity<'Key, 'Value>) =
+        lock evictionQueue
+        <| fun () ->
+
+            let node = entity.Node
+
+            // Sync between store and the eviction queue is not atomic. It might be already evicted or not yet added.
+            if node.List = evictionQueue then
+                // Just move this node to the end of the list.
+                evictionQueue.Remove(node)
+                evictionQueue.AddLast(node)
+
+    let tryEvictOne () =
+        match evictionQueue.First with
+        | null -> evictionFails.Add 1L
+        | first ->
+            match store.TryRemove(first.Value.Key) with
+            | true, removed ->
+                lock evictionQueue <| fun () -> evictionQueue.Remove(first)
+                pool.Reclaim(removed)
+                evictions.Add 1L
+            | _ -> evictionFails.Add 1L
+
+    // Non-evictable capacity.
+    let capacity = totalCapacity - headroom
+
+    let backgroundEvictionComplete = Event<_>()
+
+    let evictItems () =
+        while store.Count > capacity - headroom && evictionQueue.Count > 0 do
+            tryEvictOne ()
+
+        backgroundEvictionComplete.Trigger()
+
+    let rec backgroundEviction () =
+        async {
+            let utilization = (float store.Count / float totalCapacity)
+            // So, based on utilization this will scale the delay between 0 and 1 seconds.
+            // Worst case scenario would be when 1 second delay happens,
+            // if the cache will grow rapidly (or in bursts), it will go beyond the maximum capacity.
+            // In this case underlying dictionary will resize, AND we will have to evict items, which will likely be slow.
+            // In this case, cache stats should be used to adjust MaximumCapacity and PercentageToEvict.
+            let delay = 1000.0 - (1000.0 * utilization)
+
+            if delay > 0.0 then
+                do! Async.Sleep(int delay)
+
+            if store.Count > capacity then
+                evictItems ()
+
+            return! backgroundEviction ()
+        }
+
+    do Async.Start(backgroundEviction (), cancellationToken = cts.Token)
+
+    member val BackgroundEvictionComplete = backgroundEvictionComplete.Publish
+
+    member val Name = instanceId
+
+    member _.TryGetValue(key: 'Key, value: outref<'Value>) =
+        match store.TryGetValue(key) with
+        | true, cachedEntity ->
+            hits.Add 1L
+            Interlocked.Increment(&cachedEntity.AccessCount) |> ignore
+            updateEvictionQueue cachedEntity
```

Comment:
This means we are going into a `lock` in the critical happy path, right?

Reply:
Yes. Now that you point this out, I wonder if posting the update to a MailboxProcessor wouldn't be better.
We could remove the background eviction loop and do all the eviction related activities asynchronously in one place.